x-pack/plugins/maps/public/actions/store_actions.js

```diff
@@ -112,7 +115,7 @@ export function replaceLayerList(newLayerList) {
 
 export function addLayer(layerDescriptor) {
   return (dispatch, getState) => {
-    dispatch(clearTemporaryLayers());
+    dispatch(clearTransientLayer());
```

Comment:
I don't think we need to do embed this in here point-blank.

We get a nested flow:
- addtransientlayer
    -  wraps addLayer
        - wraps clearTransientLayer

Is it possible to keep this transient-layer state management separate?

I feel it is fairly isolated when we need to manage transient-layer state:
- clicking the add layer button: set the existing transient layer id to null
- clicking the `cancel` button: set existing transient layer id to null, remove corresponding layer from the layer-list
- clicking the save-map button: set the existing transient layer id to null, remove corresponding layer from the layer-list
- anytime on previewLayer: set the existing transient layer id to null, remove corresponding layer from the layer list, add the new layer, set the transient layer id to this new one

Maybe I'm overlooking something.

But it seems we really only need to introduce a single new action, `set_transient_layer`, similar to `set_selected_layer`. We have in place the other actions/reducers to manage the layer-list.

We can provide sugar in the action-creators, that could collate some of these groups with addLayer/removeLayer.

Reply:
The complexity usually lands in the container, the actions file or the reducer. I've made a few changes which should be in line with your suggestions and shift the complexity to the container to better isolate the workflow and eliminate needless embedded action calls.